Camera java.lang.IllegalArgumentException: Invalid image data size

expo SDK 30

randomly get exception when scan qr codes, in logcat find

```
10-15 09:44:10.482 9832-11191/? E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #8
    Process: host.exp.exponent, PID: 9832
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:353)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.IllegalArgumentException: Invalid image data size.
        at com.google.android.gms.f.b$a.a(Unknown Source:22)
        at abi30_0_0.expo.modules.barcodescanner.utils.FrameFactory.buildFrame(FrameFactory.java:13)
        at abi30_0_0.expo.modules.barcodescanner.scanners.GMVBarCodeScanner.scan(GMVBarCodeScanner.java:36)
        at abi30_0_0.expo.modules.camera.tasks.BarCodeScannerAsyncTask.doInBackground(BarCodeScannerAsyncTask.java:36)
        at abi30_0_0.expo.modules.camera.tasks.BarCodeScannerAsyncTask.doInBackground(BarCodeScannerAsyncTask.java:6)
        at android.os.AsyncTask$2.call(AsyncTask.java:333)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764) 
```

is it relased to https://github.com/react-native-community/react-native-camera/issues/1255 ?

> expo SDK 30
> 
> randomly get exception when scan qr codes, in logcat find
> 
> 10-15 09:44:10.482 9832-11191/? E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #8
> Process: host.exp.exponent, PID: 9832
> java.lang.RuntimeException: An error occurred while executing doInBackground()
> at android.os.AsyncTask$3.done(AsyncTask.java:353)
> at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
> at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
> at java.util.concurrent.FutureTask.run(FutureTask.java:271)
> at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
> at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
> at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
> at java.lang.Thread.run(Thread.java:764)
> Caused by: java.lang.IllegalArgumentException: Invalid image data size.
> at com.google.android.gms.f.b$a.a(Unknown Source:22)
> at abi30_0_0.expo.modules.barcodescanner.utils.FrameFactory.buildFrame(FrameFactory.java:13)
> at abi30_0_0.expo.modules.barcodescanner.scanners.GMVBarCodeScanner.scan(GMVBarCodeScanner.java:36)
> at abi30_0_0.expo.modules.camera.tasks.BarCodeScannerAsyncTask.doInBackground(BarCodeScannerAsyncTask.java:36)
> at abi30_0_0.expo.modules.camera.tasks.BarCodeScannerAsyncTask.doInBackground(BarCodeScannerAsyncTask.java:6)
> at android.os.AsyncTask$2.call(AsyncTask.java:333)
> at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245) 
> at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
> at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
> at java.lang.Thread.run(Thread.java:764) 
> 
> is it relased to [react-native-community/react-native-camera#1255](https://github.com/react-native-community/react-native-camera/issues/1255) ?
Could you reproduce your enviroment?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,7 @@ This is the log of notable changes to the Expo client that are developer-facing.
 - fix resumable downloads on iOS by base64-encoding `resumeData` by [@Szymon20000](https://github.com/Szymon20000) ([#2698](https://github.com/expo/expo/pull/2698))
 - fix `Permissions.LOCATION` issue that wouldn't allow asking for it in a multi-permission call by [@sjchmiela](https://github.com/sjchmiela) ([304fe560](https://github.com/expo/expo/commit/304fe560500b662be53be2c1d5a06445ad9d3702))
 - fix `onActivityResult` not being called on listeners registered to `ReactContext` by [@sjchmiela](https://github.com/sjchmiela) ([#2768](https://github.com/expo/expo/pull/2768))
+- fix fatal exception being thrown sometimes on Android when detecting barcodes by [@sjchmiela](https://github.com/sjchmiela) ([#2772](https://github.com/expo/expo/pull/2772))
 
 ## 31.0.3
 

--- a/packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/scanners/GMVBarCodeScanner.java
+++ b/packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/scanners/GMVBarCodeScanner.java
@@ -33,8 +33,18 @@ public class GMVBarCodeScanner extends ExpoBarCodeScanner {
 
   @Override
   public BarCodeScannerResult scan(byte[] data, int width, int height, int rotation) {
-    List<BarCodeScannerResult> results = scan(FrameFactory.buildFrame(data, width, height, rotation).getFrame());
-    return results.size() > 0 ? results.get(0) : null;
+    try {
+      List<BarCodeScannerResult> results = scan(FrameFactory.buildFrame(data, width, height, rotation).getFrame());
+      return results.size() > 0 ? results.get(0) : null;
+    } catch (Exception e) {
+      // Sometimes data has different size than width and height would suggest:
+      // ByteBuffer.wrap(data).capacity() < width * height.
+      // When given such arguments, Frame cannot be built and IllegalArgumentException is thrown.
+      // See https://github.com/expo/expo/issues/2422.
+      // In such case we can't do anything about it but ignore the frame.
+      Log.e(TAG, "Failed to detect barcode: " + e.getMessage());
+      return null;
+    }
   }
 
   @Override

